File: packages/rust/src/script_data/schema.rs
use indexmap::IndexMap;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// This will be used both for transliteration and typing.
///
/// Common struct for both `text_to_krama_map` and `typing_text_to_krama_map`
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct TextToKramaMap {
  pub next: Option<Vec<String>>,
  pub krama: Option<Vec<i16>>,
  pub fallback_list_ref: Option<i16>,
  /// only in `typing_text_to_krama_map`
  pub custom_back_ref: Option<i16>,
}

#[derive(Debug, Deserialize, Clone, PartialEq)]
#[serde(tag = "type")]
pub enum ListJson {
  #[serde(rename = "anya")]
  Anya { krama_ref: Vec<i16> },
  #[serde(rename = "vyanjana")]
  Vyanjana { krama_ref: Vec<i16> },
  #[serde(rename = "mAtrA")]
  Matra { krama_ref: Vec<i16> },
  #[serde(rename = "svara")]
  Svara {
    krama_ref: Vec<i16>,
    #[serde(rename = "mAtrA_krama_ref")]
    matra_krama_ref: Vec<i16>,
  },
}

#[derive(Debug, Deserialize, Serialize, Clone, PartialEq)]
pub enum List {
  Anya {
    krama_ref: Vec<i16>,
  },
  Vyanjana {
    krama_ref: Vec<i16>,
  },
  Matra {
    krama_ref: Vec<i16>,
  },
  Svara {
    krama_ref: Vec<i16>,
    matra_krama_ref: Vec<i16>,
  },
}

impl From<ListJson> for List {
  fn from(value: ListJson) -> Self {
    match value {
      ListJson::Anya { krama_ref } => List::Anya { krama_ref },
      ListJson::Vyanjana { krama_ref } => List::Vyanjana { krama_ref },
      ListJson::Matra { krama_ref } => List::Matra { krama_ref },
      ListJson::Svara {
        krama_ref,
        matra_krama_ref,
      } => List::Svara {
        krama_ref,
        matra_krama_ref,
      },
    }
  }
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct CommonScriptAttrJson {
  #[allow(dead_code)]
  pub script_name: String,
  #[allow(dead_code)]
  pub script_id: u8,
  pub krama_text_arr: Vec<(String, Option<i16>)>,
  pub krama_text_arr_index: Vec<usize>,
  pub text_to_krama_map: Vec<(String, TextToKramaMap)>,
  pub typing_text_to_krama_map: Vec<(String, TextToKramaMap)>,
  pub custom_script_chars_arr: Vec<(String, Option<i16>, Option<i16>)>,
  pub list: Vec<ListJson>,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct CommonScriptAttr {
  #[allow(dead_code)]
  pub script_name: String,
  #[allow(dead_code)]
  pub script_id: u8,
  pub krama_text_arr: Vec<(String, Option<i16>)>,
  pub krama_text_arr_index: Vec<usize>,
  pub text_to_krama_map: Vec<(String, TextToKramaMap)>,
  pub typing_text_to_krama_map: Vec<(String, TextToKramaMap)>,
  pub custom_script_chars_arr: Vec<(String, Option<i16>, Option<i16>)>,
  pub list: Vec<List>,
}

impl From<CommonScriptAttrJson> for CommonScriptAttr {
  fn from(value: CommonScriptAttrJson) -> Self {
    Self {
      script_name: value.script_name,
      script_id: value.script_id,
      krama_text_arr: value.krama_text_arr,
      krama_text_arr_index: value.krama_text_arr_index,
      text_to_krama_map: value.text_to_krama_map,
      typing_text_to_krama_map: value.typing_text_to_krama_map,
      custom_script_chars_arr: value.custom_script_chars_arr,
      list: value.list.into_iter().map(Into::into).collect(),
    }
  }
}

/// JSON representation (matches the on-disk `*.json` files).
///
/// `#[serde(flatten)]` provides a convenient shape for JSON but is not supported
/// by `bincode`, so we parse into `ScriptDataJson` and convert into `ScriptData`.
#[derive(Debug, Deserialize)]
#[serde(tag = "script_type")]
pub enum ScriptDataJson {
  #[serde(rename = "brahmic")]
  Brahmic {
    #[serde(flatten)]
    common_script_attr: CommonScriptAttrJson,
    #[allow(dead_code)]
    schwa_property: bool,
    halant: String,
    nuqta: Option<String>,
  },
  #[serde(rename = "other")]
  Other {
    #[serde(flatten)]
    common_script_attr: CommonScriptAttrJson,
    #[allow(dead_code)]
    schwa_character: String,
  },
}

#[derive(Debug, Deserialize, Serialize)]
pub enum ScriptData {
  Brahmic {
    common_script_attr: CommonScriptAttr,
    #[allow(dead_code)]
    schwa_property: bool,
    halant: String,
    nuqta: Option<String>,
  },
  Other {
    common_script_attr: CommonScriptAttr,
    #[allow(dead_code)]
    schwa_character: String,
  },
}

impl From<ScriptDataJson> for ScriptData {
  fn from(value: ScriptDataJson) -> Self {
    match value {
      ScriptDataJson::Brahmic {
        common_script_attr,
        schwa_property,
        halant,
        nuqta,
      } => ScriptData::Brahmic {
        common_script_attr: common_script_attr.into(),
        schwa_property,
        halant,
        nuqta,
      },
      ScriptDataJson::Other {
        common_script_attr,
        schwa_character,
      } => ScriptData::Other {
        common_script_attr: common_script_attr.into(),
        schwa_character,
      },
    }
  }
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum CustomOptionScriptTypeEnum {
  Brahmic,
  Other,
  All,
}

#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum CheckInEnum {
  Input,
  Output,
}

#[derive(Debug, Deserialize)]
#[serde(tag = "type")]
pub enum RuleJson {
  #[serde(rename = "replace_prev_krama_keys")]
  ReplacePrevKramaKeys {
    #[serde(default)]
    use_replace: Option<bool>,
    prev: Vec<i16>,
    following: Vec<i16>,
    replace_with: Vec<i16>,
    check_in: Option<CheckInEnum>,
  },
  #[serde(rename = "direct_replace")]
  DirectReplace {
    #[serde(default)]
    use_replace: Option<bool>,
    to_replace: Vec<Vec<i16>>,
    replace_with: Vec<i16>,
    replace_text: Option<String>,
    check_in: Option<CheckInEnum>,
  },
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub enum Rule {
  ReplacePrevKramaKeys {
    use_replace: Option<bool>,
    prev: Vec<i16>,
    following: Vec<i16>,
    replace_with: Vec<i16>,
    check_in: Option<CheckInEnum>,
  },
  DirectReplace {
    use_replace: Option<bool>,
    to_replace: Vec<Vec<i16>>,
    replace_with: Vec<i16>,
    replace_text: Option<String>,
    check_in: Option<CheckInEnum>,
  },
}

impl From<RuleJson> for Rule {
  fn from(value: RuleJson) -> Self {
    match value {
      RuleJson::ReplacePrevKramaKeys {
        use_replace,
        prev,
        following,
        replace_with,
        check_in,
      } => Rule::ReplacePrevKramaKeys {
        use_replace,
        prev,
        following,
        replace_with,
        check_in,
      },
      RuleJson::DirectReplace {
        use_replace,
        to_replace,
        replace_with,
        replace_text,
        check_in,
      } => Rule::DirectReplace {
        use_replace,
        to_replace,
        replace_with,
        replace_text,
        check_in,
      },
    }
  }
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct CustomOptionsJson {
  pub from_script_name: Option<Vec<String>>,
  pub from_script_type: Option<CustomOptionScriptTypeEnum>,
  pub to_script_name: Option<Vec<String>>,
  pub to_script_type: Option<CustomOptionScriptTypeEnum>,
  #[allow(dead_code)]
  pub check_in: CheckInEnum,
  pub rules: Vec<RuleJson>,
}

impl From<CustomOptionsJson> for CustomOptions {
  fn from(value: CustomOptionsJson) -> Self {
    Self {
      from_script_name: value.from_script_name,
      from_script_type: value.from_script_type,
      to_script_name: value.to_script_name,
      to_script_type: value.to_script_type,
      check_in: value.check_in,
      rules: value.rules.into_iter().map(Into::into).collect(),
    }
  }
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct CustomOptions {
  pub from_script_name: Option<Vec<String>>,
  pub from_script_type: Option<CustomOptionScriptTypeEnum>,
  pub to_script_name: Option<Vec<String>>,
  pub to_script_type: Option<CustomOptionScriptTypeEnum>,
  #[allow(dead_code)]
  pub check_in: CheckInEnum,
  pub rules: Vec<Rule>,
}

#[allow(dead_code)]
pub type CustomOptionMapJson = IndexMap<String, CustomOptionsJson>;
pub type CustomOptionMap = IndexMap<String, CustomOptions>;

#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct ScriptListDataJson {
  pub scripts: HashMap<String, u8>,
  pub langs: HashMap<String, u8>,
  /// all langs are mapped to a script
  pub lang_script_map: HashMap<String, String>,
  /// contains aliases which map to script
  pub script_alternates_map: HashMap<String, String>,
}

File: packages/zig/src/script_data/schema.zig
const std = @import("std");

pub const Allocator = std.mem.Allocator;
pub const JsonValue = std.json.Value;
pub const JsonObjectMap = std.json.ObjectMap;
pub const JsonArray = std.json.Array;

pub const ParseError = error{
    InvalidSchema,
    MissingField,
    UnsupportedNumber,
};

pub const TextToKramaMap = struct {
    next: ?[]const []const u8,
    krama: ?[]const i16,
    fallback_list_ref: ?i16,
    custom_back_ref: ?i16,
};

pub const KramaTextEntry = struct {
    text: []const u8,
    list_arr_ref: ?i16,
};

pub const TextMapEntry = struct {
    text: []const u8,
    value: TextToKramaMap,
};

pub const CustomScriptCharEntry = struct {
    text: []const u8,
    primary_ref: ?i16,
    secondary_ref: ?i16,
};

pub const SimpleList = struct {
    krama_ref: []const i16,
};

pub const SvaraList = struct {
    krama_ref: []const i16,
    matra_krama_ref: []const i16,
};

pub const List = union(enum) {
    anya: SimpleList,
    vyanjana: SimpleList,
    @"mAtrA": SimpleList,
    svara: SvaraList,

    pub fn getKramaRef(self: List) []const i16 {
        return switch (self) {
            .anya => |value| value.krama_ref,
            .vyanjana => |value| value.krama_ref,
            .@"mAtrA" => |value| value.krama_ref,
            .svara => |value| value.krama_ref,
        };
    }

    pub fn isSvara(self: List) bool {
        return self == .svara;
    }

    pub fn isMatra(self: List) bool {
        return self == .@"mAtrA";
    }

    pub fn isVyanjana(self: List) bool {
        return self == .vyanjana;
    }

    pub fn isAnya(self: List) bool {
        return self == .anya;
    }
};

pub const CommonScriptAttr = struct {
    script_name: []const u8,
    script_id: u8,
    krama_text_arr: []const KramaTextEntry,
    krama_text_arr_index: []const usize,
    text_to_krama_map: []const TextMapEntry,
    typing_text_to_krama_map: []const TextMapEntry,
    custom_script_chars_arr: []const CustomScriptCharEntry,
    list: []const List,

    krama_text_lookup: std.StringHashMapUnmanaged(usize) = .{},
    text_to_krama_lookup: std.StringHashMapUnmanaged(usize) = .{},
    typing_text_to_krama_lookup: std.StringHashMapUnmanaged(usize) = .{},
    custom_script_chars_lookup: std.StringHashMapUnmanaged(usize) = .{},

    pub fn initLookups(self: *CommonScriptAttr, allocator: Allocator) !void {
        try self.krama_text_lookup.ensureTotalCapacity(allocator, @intCast(self.krama_text_arr.len));
        for (self.krama_text_arr, 0..) |entry, index| {
            const gop = try self.krama_text_lookup.getOrPut(allocator, entry.text);
            if (!gop.found_existing) {
                gop.value_ptr.* = index;
            }
        }

        try self.text_to_krama_lookup.ensureTotalCapacity(allocator, @intCast(self.text_to_krama_map.len));
        for (self.text_to_krama_map, 0..) |entry, index| {
            const gop = try self.text_to_krama_lookup.getOrPut(allocator, entry.text);
            if (!gop.found_existing) {
                gop.value_ptr.* = index;
            }
        }

        try self.typing_text_to_krama_lookup.ensureTotalCapacity(allocator, @intCast(self.typing_text_to_krama_map.len));
        for (self.typing_text_to_krama_map, 0..) |entry, index| {
            const gop = try self.typing_text_to_krama_lookup.getOrPut(allocator, entry.text);
            if (!gop.found_existing) {
                gop.value_ptr.* = index;
            }
        }

        try self.custom_script_chars_lookup.ensureTotalCapacity(allocator, @intCast(self.custom_script_chars_arr.len));
        for (self.custom_script_chars_arr, 0..) |entry, index| {
            const gop = try self.custom_script_chars_lookup.getOrPut(allocator, entry.text);
            if (!gop.found_existing) {
                gop.value_ptr.* = index;
            }
        }
    }
};

pub const BrahmicScriptData = struct {
    common_script_attr: CommonScriptAttr,
    schwa_property: bool,
    halant: []const u8,
    nuqta: ?[]const u8,
};

pub const OtherScriptData = struct {
    common_script_attr: CommonScriptAttr,
    schwa_character: []const u8,
};

pub const ScriptData = union(enum) {
    brahmic: BrahmicScriptData,
    other: OtherScriptData,

    pub fn getCommonAttr(self: *const ScriptData) *const CommonScriptAttr {
        return switch (self.*) {
            .brahmic => |*value| &value.common_script_attr,
            .other => |*value| &value.common_script_attr,
        };
    }

    pub fn getCommonAttrMut(self: *ScriptData) *CommonScriptAttr {
        return switch (self.*) {
            .brahmic => |*value| &value.common_script_attr,
            .other => |*value| &value.common_script_attr,
        };
    }

    pub fn scriptName(self: *const ScriptData) []const u8 {
        return self.getCommonAttr().script_name;
    }

    pub fn scriptType(self: *const ScriptData) CustomOptionScriptType {
        return switch (self.*) {
            .brahmic => .brahmic,
            .other => .other,
        };
    }

    pub fn initLookups(self: *ScriptData, allocator: Allocator) !void {
        try self.getCommonAttrMut().initLookups(allocator);
    }

    pub fn textToKramaMapIndex(self: *const ScriptData, text: []const u8, use_typing_map: bool) ?usize {
        const attr = self.getCommonAttr();
        if (use_typing_map) {
            return attr.typing_text_to_krama_lookup.get(text);
        }
        return attr.text_to_krama_lookup.get(text);
    }

    pub fn customScriptCharIndexOfText(self: *const ScriptData, text: []const u8) ?usize {
        return self.getCommonAttr().custom_script_chars_lookup.get(text);
    }
};

pub const CustomOptionScriptType = enum {
    brahmic,
    other,
    all,
};

pub const CheckIn = enum {
    input,
    output,
};

pub const ReplacePrevKramaKeysRule = struct {
    use_replace: ?bool,
    prev: []const i16,
    following: []const i16,
    replace_with: []const i16,
    check_in: ?CheckIn,
};

pub const DirectReplaceRule = struct {
    use_replace: ?bool,
    to_replace: []const []const i16,
    replace_with: []const i16,
    replace_text: ?[]const u8,
    check_in: ?CheckIn,
};

pub const Rule = union(enum) {
    replace_prev_krama_keys: ReplacePrevKramaKeysRule,
    direct_replace: DirectReplaceRule,
};

pub const CustomOption = struct {
    from_script_name: ?[]const []const u8,
    from_script_type: ?CustomOptionScriptType,
    to_script_name: ?[]const []const u8,
    to_script_type: ?CustomOptionScriptType,
    check_in: CheckIn,
    rules: []const Rule,
};

pub fn parseScriptDataJsonBytes(allocator: Allocator, bytes: []const u8) !ScriptData {
    const parsed = try std.json.parseFromSlice(JsonValue, allocator, bytes, .{});
    return parseScriptDataValue(allocator, parsed.value);
}

pub fn parseScriptDataValue(allocator: Allocator, value: JsonValue) !ScriptData {
    const object = try expectObject(value);
    const script_type = try getRequiredString(object, "script_type");
    const common = try parseCommonScriptAttr(allocator, object);

    if (std.mem.eql(u8, script_type, "brahmic")) {
        return .{
            .brahmic = .{
                .common_script_attr = common,
                .schwa_property = try getRequiredBool(object, "schwa_property"),
                .halant = try getRequiredString(object, "halant"),
                .nuqta = try getOptionalString(object, "nuqta"),
            },
        };
    }

    if (std.mem.eql(u8, script_type, "other")) {
        return .{
            .other = .{
                .common_script_attr = common,
                .schwa_character = try getRequiredString(object, "schwa_character"),
            },
        };
    }

    return error.InvalidSchema;
}

pub fn parseCommonScriptAttr(allocator: Allocator, object: JsonObjectMap) !CommonScriptAttr {
    return .{
        .script_name = try getRequiredString(object, "script_name"),
        .script_id = try parseU8(try getRequiredValue(object, "script_id")),
        .krama_text_arr = try parseKramaTextArray(allocator, try getRequiredValue(object, "krama_text_arr")),
        .krama_text_arr_index = try parseUsizeArray(allocator, try getRequiredValue(object, "krama_text_arr_index")),
        .text_to_krama_map = try parseTextMapEntries(allocator, try getRequiredValue(object, "text_to_krama_map")),
        .typing_text_to_krama_map = try parseTextMapEntries(allocator, try getRequiredValue(object, "typing_text_to_krama_map")),
        .custom_script_chars_arr = try parseCustomScriptCharsArray(allocator, try getRequiredValue(object, "custom_script_chars_arr")),
        .list = try parseListArray(allocator, try getRequiredValue(object, "list")),
    };
}

pub fn parseTextToKramaMap(allocator: Allocator, value: JsonValue) !TextToKramaMap {
    const object = try expectObject(value);
    return .{
        .next = try parseOptionalStringArrayField(allocator, object, "next"),
        .krama = try parseOptionalI16ArrayField(allocator, object, "krama"),
        .fallback_list_ref = try parseOptionalI16Field(object, "fallback_list_ref"),
        .custom_back_ref = try parseOptionalI16Field(object, "custom_back_ref"),
    };
}

pub fn parseList(allocator: Allocator, value: JsonValue) !List {
    const object = try expectObject(value);
    const list_type = try getRequiredString(object, "type");
    const krama_ref = try parseI16Array(allocator, try getRequiredValue(object, "krama_ref"));

    if (std.mem.eql(u8, list_type, "anya")) {
        return .{ .anya = .{ .krama_ref = krama_ref } };
    }
    if (std.mem.eql(u8, list_type, "vyanjana")) {
        return .{ .vyanjana = .{ .krama_ref = krama_ref } };
    }
    if (std.mem.eql(u8, list_type, "mAtrA")) {
        return .{ .@"mAtrA" = .{ .krama_ref = krama_ref } };
    }
    if (std.mem.eql(u8, list_type, "svara")) {
        return .{
            .svara = .{
                .krama_ref = krama_ref,
                .matra_krama_ref = try parseI16Array(allocator, try getRequiredValue(object, "mAtrA_krama_ref")),
            },
        };
    }
    return error.InvalidSchema;
}

pub fn parseCustomOptionScriptType(value: JsonValue) !CustomOptionScriptType {
    const string_value = try expectString(value);
    if (std.mem.eql(u8, string_value, "brahmic")) return .brahmic;
    if (std.mem.eql(u8, string_value, "other")) return .other;
    if (std.mem.eql(u8, string_value, "all")) return .all;
    return error.InvalidSchema;
}

pub fn parseCheckIn(value: JsonValue) !CheckIn {
    const string_value = try expectString(value);
    if (std.mem.eql(u8, string_value, "input")) return .input;
    if (std.mem.eql(u8, string_value, "output")) return .output;
    return error.InvalidSchema;
}

pub fn parseRule(allocator: Allocator, value: JsonValue) !Rule {
    const object = try expectObject(value);
    const rule_type = try getRequiredString(object, "type");
    if (std.mem.eql(u8, rule_type, "replace_prev_krama_keys")) {
        return .{
            .replace_prev_krama_keys = .{
                .use_replace = try parseOptionalBoolField(object, "use_replace"),
                .prev = try parseI16Array(allocator, try getRequiredValue(object, "prev")),
                .following = try parseI16Array(allocator, try getRequiredValue(object, "following")),
                .replace_with = try parseI16Array(allocator, try getRequiredValue(object, "replace_with")),
                .check_in = try parseOptionalCheckInField(object, "check_in"),
            },
        };
    }

    if (std.mem.eql(u8, rule_type, "direct_replace")) {
        return .{
            .direct_replace = .{
                .use_replace = try parseOptionalBoolField(object, "use_replace"),
                .to_replace = try parseArrayOfI16Array(allocator, try getRequiredValue(object, "to_replace")),
                .replace_with = try parseI16Array(allocator, try getRequiredValue(object, "replace_with")),
                .replace_text = try getOptionalString(object, "replace_text"),
                .check_in = try parseOptionalCheckInField(object, "check_in"),
            },
        };
    }

    return error.InvalidSchema;
}

pub fn parseCustomOption(allocator: Allocator, value: JsonValue) !CustomOption {
    const object = try expectObject(value);
    return .{
        .from_script_name = try parseOptionalStringArrayField(allocator, object, "from_script_name"),
        .from_script_type = try parseOptionalScriptTypeField(object, "from_script_type"),
        .to_script_name = try parseOptionalStringArrayField(allocator, object, "to_script_name"),
        .to_script_type = try parseOptionalScriptTypeField(object, "to_script_type"),
        .check_in = try parseCheckIn(try getRequiredValue(object, "check_in")),
        .rules = try parseRulesArray(allocator, try getRequiredValue(object, "rules")),
    };
}

pub fn expectObject(value: JsonValue) !JsonObjectMap {
    return switch (value) {
        .object => |object| object,
        else => error.InvalidSchema,
    };
}

pub fn expectArray(value: JsonValue) !JsonArray {
    return switch (value) {
        .array => |array| array,
        else => error.InvalidSchema,
    };
}

pub fn expectString(value: JsonValue) ![]const u8 {
    return switch (value) {
        .string => |string_value| string_value,
        else => error.InvalidSchema,
    };
}

pub fn expectBool(value: JsonValue) !bool {
    return switch (value) {
        .bool => |bool_value| bool_value,
        else => error.InvalidSchema,
    };
}

pub fn parseI16(value: JsonValue) !i16 {
    return switch (value) {
        .integer => |integer_value| std.math.cast(i16, integer_value) orelse error.UnsupportedNumber,
        else => error.InvalidSchema,
    };
}

pub fn parseU8(value: JsonValue) !u8 {
    return switch (value) {
        .integer => |integer_value| std.math.cast(u8, integer_value) orelse error.UnsupportedNumber,
        else => error.InvalidSchema,
    };
}

pub fn parseUsize(value: JsonValue) !usize {
    return switch (value) {
        .integer => |integer_value| std.math.cast(usize, integer_value) orelse error.UnsupportedNumber,
        else => error.InvalidSchema,
    };
}

pub fn getRequiredValue(object: JsonObjectMap, key: []const u8) !JsonValue {
    return object.get(key) orelse error.MissingField;
}

pub fn getRequiredString(object: JsonObjectMap, key: []const u8) ![]const u8 {
    return expectString(try getRequiredValue(object, key));
}

pub fn getRequiredBool(object: JsonObjectMap, key: []const u8) !bool {
    return expectBool(try getRequiredValue(object, key));
}

pub fn getOptionalString(object: JsonObjectMap, key: []const u8) !?[]const u8 {
    const value = object.get(key) orelse return null;
    return switch (value) {
        .null => null,
        .string => |string_value| string_value,
        else => error.InvalidSchema,
    };
}

pub fn parseOptionalBoolField(object: JsonObjectMap, key: []const u8) !?bool {
    const value = object.get(key) orelse return null;
    return switch (value) {
        .null => null,
        .bool => |bool_value| bool_value,
        else => error.InvalidSchema,
    };
}

pub fn parseOptionalI16Field(object: JsonObjectMap, key: []const u8) !?i16 {
    const value = object.get(key) orelse return null;
    return switch (value) {
        .null => null,
        else => try parseI16(value),
    };
}

pub fn parseOptionalScriptTypeField(object: JsonObjectMap, key: []const u8) !?CustomOptionScriptType {
    const value = object.get(key) orelse return null;
    return switch (value) {
        .null => null,
        else => try parseCustomOptionScriptType(value),
    };
}

pub fn parseOptionalCheckInField(object: JsonObjectMap, key: []const u8) !?CheckIn {
    const value = object.get(key) orelse return null;
    return switch (value) {
        .null => null,
        else => try parseCheckIn(value),
    };
}

pub fn parseOptionalStringArrayField(allocator: Allocator, object: JsonObjectMap, key: []const u8) !?[]const []const u8 {
    const value = object.get(key) orelse return null;
    return switch (value) {
        .null => null,
        else => try parseStringArray(allocator, value),
    };
}

pub fn parseOptionalI16ArrayField(allocator: Allocator, object: JsonObjectMap, key: []const u8) !?[]const i16 {
    const value = object.get(key) orelse return null;
    return switch (value) {
        .null => null,
        else => try parseI16Array(allocator, value),
    };
}

pub fn parseStringArray(allocator: Allocator, value: JsonValue) ![]const []const u8 {
    const array = try expectArray(value);
    const slice = try allocator.alloc([]const u8, array.items.len);
    for (array.items, 0..) |item, index| {
        slice[index] = try expectString(item);
    }
    return slice;
}

pub fn parseI16Array(allocator: Allocator, value: JsonValue) ![]const i16 {
    const array = try expectArray(value);
    const slice = try allocator.alloc(i16, array.items.len);
    for (array.items, 0..) |item, index| {
        slice[index] = try parseI16(item);
    }
    return slice;
}

pub fn parseUsizeArray(allocator: Allocator, value: JsonValue) ![]const usize {
    const array = try expectArray(value);
    const slice = try allocator.alloc(usize, array.items.len);
    for (array.items, 0..) |item, index| {
        slice[index] = try parseUsize(item);
    }
    return slice;
}

pub fn parseArrayOfI16Array(allocator: Allocator, value: JsonValue) ![]const []const i16 {
    const array = try expectArray(value);
    const slice = try allocator.alloc([]const i16, array.items.len);
    for (array.items, 0..) |item, index| {
        slice[index] = try parseI16Array(allocator, item);
    }
    return slice;
}

pub fn parseKramaTextArray(allocator: Allocator, value: JsonValue) ![]const KramaTextEntry {
    const array = try expectArray(value);
    const slice = try allocator.alloc(KramaTextEntry, array.items.len);
    for (array.items, 0..) |item, index| {
        const tuple = try expectArray(item);
        if (tuple.items.len != 2) return error.InvalidSchema;
        slice[index] = .{
            .text = try expectString(tuple.items[0]),
            .list_arr_ref = switch (tuple.items[1]) {
                .null => null,
                else => try parseI16(tuple.items[1]),
            },
        };
    }
    return slice;
}

pub fn parseTextMapEntries(allocator: Allocator, value: JsonValue) ![]const TextMapEntry {
    const array = try expectArray(value);
    const slice = try allocator.alloc(TextMapEntry, array.items.len);
    for (array.items, 0..) |item, index| {
        const tuple = try expectArray(item);
        if (tuple.items.len != 2) return error.InvalidSchema;
        slice[index] = .{
            .text = try expectString(tuple.items[0]),
            .value = try parseTextToKramaMap(allocator, tuple.items[1]),
        };
    }
    return slice;
}

pub fn parseCustomScriptCharsArray(allocator: Allocator, value: JsonValue) ![]const CustomScriptCharEntry {
    const array = try expectArray(value);
    const slice = try allocator.alloc(CustomScriptCharEntry, array.items.len);
    for (array.items, 0..) |item, index| {
        const tuple = try expectArray(item);
        if (tuple.items.len != 3) return error.InvalidSchema;
        slice[index] = .{
            .text = try expectString(tuple.items[0]),
            .primary_ref = switch (tuple.items[1]) {
                .null => null,
                else => try parseI16(tuple.items[1]),
            },
            .secondary_ref = switch (tuple.items[2]) {
                .null => null,
                else => try parseI16(tuple.items[2]),
            },
        };
    }
    return slice;
}

pub fn parseListArray(allocator: Allocator, value: JsonValue) ![]const List {
    const array = try expectArray(value);
    const slice = try allocator.alloc(List, array.items.len);
    for (array.items, 0..) |item, index| {
        slice[index] = try parseList(allocator, item);
    }
    return slice;
}

pub fn parseRulesArray(allocator: Allocator, value: JsonValue) ![]const Rule {
    const array = try expectArray(value);
    const slice = try allocator.alloc(Rule, array.items.len);
    for (array.items, 0..) |item, index| {
        slice[index] = try parseRule(allocator, item);
    }
    return slice;
}
